src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,192 @@
+    [Shared]
+    [Export(typeof(LSPProgressListener))]
+    internal class DefaultLSPProgressListener : LSPProgressListener, IDisposable
+    {
+        private const string ProgressNotificationValueName = "value";
+
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+
+        private readonly ConcurrentDictionary<string, ProgressRequest> _activeRequests
+            = new ConcurrentDictionary<string, ProgressRequest>();
+
+        [ImportingConstructor]
+        public DefaultLSPProgressListener(ILanguageServiceBroker2 languageServiceBroker)
+        {
+            if (languageServiceBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageServiceBroker));
+            }
+
+            _languageServiceBroker = languageServiceBroker;
+
+            _languageServiceBroker.ClientNotifyAsync += ClientNotifyAsyncListenerAsync;
+        }
+
+        private Task ClientNotifyAsyncListenerAsync(object sender, LanguageClientNotifyEventArgs args)
+        {
+            return ProcessProgressNotificationAsync(args.MethodName, args.ParameterToken);
+        }
+
+        private async Task ProcessProgressNotificationAsync(string methodName, JToken parameterToken)
+        {
+            if (methodName != Methods.ProgressNotificationName ||
+               !parameterToken.HasValues ||
+               parameterToken[ProgressNotificationValueName] is null ||
+               parameterToken[Methods.ProgressNotificationTokenName] is null)
+            {
+                return;
+            }
+
+            var token = parameterToken[Methods.ProgressNotificationTokenName].ToObject<string>(); // IProgress<object>>();
+
+            if (string.IsNullOrEmpty(token) || !_activeRequests.TryGetValue(token, out var request))
+            {
+                return;
+            }
+
+            var value = parameterToken[ProgressNotificationValueName];
+
+            try
+            {
+                request.HandlerCancellationToken.ThrowIfCancellationRequested();
+                await request.OnProgressNotifyAsync(value).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Initial handle request has been cancelled
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            CompleteAfterDelay(token, request);
+        }
+
+        private void CompleteAfterDelay(string token, ProgressRequest request)
+        {
+            CancellationTokenSource linkedCTS;
+
+            lock (request.RequestLock)
+            {
+                var existingTimeoutCTS = request.TimeoutCancellationTokenSource;
+                if (existingTimeoutCTS != null &&
+                    existingTimeoutCTS.Token.CanBeCanceled &&
+                    !existingTimeoutCTS.Token.IsCancellationRequested)
+                {
+                    existingTimeoutCTS.Cancel();
```

Comment:
I believe cts' also need to have their Dispose method called when they're "done with"